src/main/java/org/edgexfoundry/messaging/ZeroMQEventSubscriber.java

```diff
@@ -33,119 +35,126 @@
-      subscriber.close();
-    subscriber = null;
-    // try to restart
-    logger.debug("Attempting restart of Event message watch.");
-    receive();
-  }
-
-  public String getZeromqAddress() {
-    return zeromqAddress;
-  }
-
-  public void setZeromqAddress(String zeromqAddress) {
-    this.zeromqAddress = zeromqAddress;
-  }
-
-  public String getZeromqAddressPort() {
-    return zeromqAddressPort;
-  }
-
-  public void setZeromqAddressPort(String zeromqAddressPort) {
-    this.zeromqAddressPort = zeromqAddressPort;
-  }
-
-  private ZMQ.Socket getSubscriber() {
-    if (subscriber == null) {
-      try {
-        subscriber = context.socket(ZMQ.SUB);
-        subscriber.connect(zeromqAddress + ":" + zeromqAddressPort);
-        subscriber.subscribe("".getBytes());
-      } catch (Exception e) {
-        logger.error("Unable to get a ZMQ subscriber.  Error:  " + e);
-        subscriber = null;
-      }
-    }
-    return subscriber;
-  }
-
-  private Event toEvent(String eventString) throws IOException, ClassNotFoundException {
-    byte[] data = Base64.getDecoder().decode(eventString);
-    ObjectInputStream in = new ObjectInputStream(new ByteArrayInputStream(data));
-    Event event = (Event) in.readObject();
-    in.close();
-    return event;
-  }
-
-  private static Event toEvent(byte[] eventBytes) throws IOException, ClassNotFoundException {
-    try {
-      Gson gson = new Gson();
-      String json = new String(eventBytes);
-      return gson.fromJson(json, Event.class);
-    } catch (Exception e) {
-      // Try to degrade to deprecated serialization functionality gracefully
-      ByteArrayInputStream bis = new ByteArrayInputStream(eventBytes);
-      ObjectInput in = new ObjectInputStream(bis);
-      Event event = (Event) in.readObject();
-      return event;
-    }
-  }
+	private static final org.edgexfoundry.support.logging.client.EdgeXLogger logger = org.edgexfoundry.support.logging.client.EdgeXLoggerFactory
+			.getEdgeXLogger(ZeroMQEventSubscriber.class);
+
+	@Value("${export.zeromq.port}")
+	private String zeromqAddressPort;
+	@Value("${export.zeromq.host}")
+	private String zeromqAddress;
+	@Value("${export.client}")
+	private boolean exportClient;
+	@Value("${expect.serializedjava}")
+	private boolean serializedJava;
+
+	@Autowired
+	RuleEngine engine;
+
+	private ZMQ.Socket subscriber;
+	private ZMQ.Context context;
+
+	{
+		context = ZMQ.context(1);
+	}
+
+	public void receive() {
+		getSubscriber();
+		String exportString = null;
+		byte[] exportBytes = null;
+		Event event;
+		logger.info("Watching for new exported Event messages...");
+		try {
+			while (!Thread.currentThread().isInterrupted()) {
+				if (exportClient) {
+					if (serializedJava) { // supporting legacy Java export distro that shipped Serialized eventString
+						exportString = subscriber.recvStr();
+						event = toEvent(exportString);
+					} else { // supporting new Go export distro that ships JSON
+						exportBytes = subscriber.recv();
```

Comment:
Seems like a decent functional start.
For more deterministic handling, is it feasible to access/deserialize the [ContentType ](https://github.com/edgexfoundry/go-mod-messaging/blob/master/pkg/types/message_envelope.go#L38) property of MessageEnvelope?

Reply:
Toby, we would need a non-binary message envelop that would tip off Java or other system of what is contained in the envelop.  Right now, everything is encoded in binary - so we'd have to fail the JSON unmarshalling and then decode the binary just to get the type.  That's a lot of work for no gain at this point in time.  Going forward, I would really like to explore an envelop that itself is not encoded/binary.

Reply:
Or possibly a message bus that supports the use of headers on messages to capture this kind of metadata.